### ISSUE TITLE
OWLQN-B

### ISSUE THREAD

**Issue by dbtsai** [scalanlp/breeze#455]
I saw that LBFGS-B was implemented couple month ago, https://github.com/scalanlp/breeze/blame/master/math/src/main/scala/breeze/optimize/LBFGSB.scala

I planed to implement Robust Regression with Elastic-Net based on Prof. Art Owen's paper, [A robust hybrid of lasso and ridge regression](http://statweb.stanford.edu/~owen/reports/hhu.pdf ).

The scaling factor in Eq.(6) \sigma has to be >= 0, so it becomes bound problem. Any plan to add this feature to OWLQN for L1? 

Thanks.

**Comment by dlwh** [comment]
No plan at all... i'd happily take such a thing of course, but I don't think it would be anywhere near easy, unless you have a cite...

For just maintaining positivity, set \sigma = exp(\tau), which is what you'd probably get if you did Lagrange multipliers, anyhow.

**Comment by dbtsai** [comment]
@MechCoder tried the `exp` trick but it seems that it's not working properly. He did it in python scikit-learn, and I will try the same thing in Spark and report back. Thanks.
https://github.com/scikit-learn/scikit-learn/pull/5291#issuecomment-147613323

**Comment by dlwh** [comment]
hrm, is there something that makes it numerically unstable with the exp
thing? it's usually pretty safe in my (limited) experience. You could also
try throwing PQN at it.

**Comment by dbtsai** [comment]
Do you think simply replacing `\sigma = exp(\tau)` will still have the loss convex? @MechCoder said he can not prove it without a fair amount of symbol crunching.

**Comment by dlwh** [comment]
http://www.wolframalpha.com/input/?i=D%5BD%5Bpiecewise%5B%7B%7Bexp%28x%29%5E2%2C+exp%28x%29+%3C%3D+1%7D%2C+%7B2+*+exp%28x%29+-+1%2C+%7Cexp%28x%29%7C+%3E%3D+1%7D%7D%5D%2C+x%5D%2C+x%5D

second derivative is positive for all x

**Comment by dbtsai** [comment]
Cool. Let's me give it a shot with `exp` trick. Thanks.

**Comment by MechCoder** [comment]
Hi, I was not talking about the loss function. I was talking about proving the joint convexity in Eq 8, when `sigma` is replaced by `exp(sigma)`

**Comment by MechCoder** [comment]
Compare the differences.

Here, `x = residual term`, and `y = sigma`. We parametrize `x` and `y` as `xo + c*t` and `y0 + st`. The goal is to prove that the double derivative is positive for positive values of t and all values of c and s.

http://www.wolframalpha.com/input/?i=D[D[piecewise[{{%28x0+%2B+ct%29**2%2F+%28y0+%2B+ct%29%2C+|x0+%2B+ct|+%3C%3D+M*%28y0+%2B+ct%29}%2C+{2+*+M+*+|x0+%2B+ct|+-+M**2*%28y0+%2B+ct%29%2C+|x0+%2B+ct|+%3EM*%28y0%2Bct%29}}]%2C+t]%2C+t]

The first term is positive for all values of c and s, while the second term is zero.

While we replace `y` with `exp(y)`
http://www.wolframalpha.com/input/?i=D[D[piecewise[{{%28x0+%2B+ct%29**2*exp%28-%28y0+%2B+ct%29%29%2C+|x0+%2B+ct|+%3C%3D+M*exp%28y0+%2B+ct%29}%2C+{2+*+M+*+|x0+%2B+ct|+-+M**2*exp%28y0+%2B+ct%29%2C+|x0+%2B+ct|+%3EM*exp%28y0%2Bct%29}}]%2C+t]%2C+t]

It does not seem so straightforward to prove.

**Comment by dlwh** [comment]
fair enough. I'll think on it later.

All that said, @dbtsai, i think you can use the fact that l1 regularization
can be expressed as box constraints, see, e.g., equation 8 in
http://pages.cs.wisc.edu/~gfung/GeneralL1/FastGeneralL1.pdf at the cost of
doubling the number of parameters (perhaps you guys already considered
this...)

If doubling the number of parameters ends up being a problem, we can
probably enable sparsity in LBFGS-B.

-- David

**Comment by dlwh** [comment]
alternatively, we can just look at

http://www.wolframalpha.com/input/?i=exp%28x%29+*+piecewise%5B%7B%7Bexp%28x%29%5E-2%2C+exp%28-x%29+%3C%3D+1%7D%2C+%7B2+%2F+exp%28x%29+-+1%2C+exp%28-x%29+%3E%3D+1%7D%7D%5D

which is clearly only convex for >= 0.

So bound constraints it is!

**Comment by MechCoder** [comment]
yeah :P I just got carried away.

**Comment by dlwh** [comment]
given the bounded formulation of L1 and the fact that a true OWLQN-B is (somewhat) novel research, I'm gonna close this...
